kip-0028.md

```diff
@@ -0,0 +1,174 @@
+name `Y` and its containing module `N` are not determined by the proof, i.e. `subj`; furthermore, there is no nonce preventing replays. In general, SPV proofs are valid to use in any module, on any chain, and are legal to use multiple times. Users can add nonces and chain IDs to their event arguments if they wish to change that for their use-case.
+
+This verifier subsumes the `verify-spv` function in Pact, though with a drastically different interface.
+
+### Hyperlane Message
+
+We define the **Hyperlane Message** verifier as follows, though this is an abstract definition without all of the nuance of the Hyperlane protocol. Its use is to verify signatures of cross-network Hyperlane messages.
+
+Given:
+  - a bridge module `B`
+  - token module `M`
+  - a recipient account `recipient`
+  - an amount `amt`
+  - a list of signing keypairs `(signerPublicKeys, signerPrivateKeys)`
+  - a Hyperlane message `msg`, which is an encoded Hyperlane v3 message as detailed in https://docs.hyperlane.xyz/docs/reference/libraries/message, which:
+    - contains as its payload the triple `(M, recipient, amt)`
+    - which as its recipient address an encoded form of `B`
+    - which is signed by `signerPrivateKeys`
+  - the message ID `messageId` that is the keccak256 hash of `msg`:
+
+The **Hyperlane Message** verifier has as input `msg` and as output the singleton set containing the Pact capability `(B.MINT M signerPublicKeys messageId amt recipient)`. The bridge module is responsible for:
+  - registering `messageId` as received in its storage, preventing message replays;
+  - checking that signature by `signerPublicKeys` is sufficient to authenticate the message;
+  - calling into `M` to ensure that the mint happens.
+
+Conventionally, Hyperlane bridges are secured by "Interchain Security Modules" or ISMs, which both contain cryptographic code and manage the set of *validators*, that is, public keys that can authenticate Hyperlane messages. This verifier implements the *mechanism* of the ISM, that is the cryptography, while leaving the *policy* of the ISM up to the bridge code, that is, managing the set of validators and ensuring that the signers match the validators in some way.
+
+# Implementation
+
+## `chainweb-node`
+
+In `chainweb-node` we propose adding a mapping from verifier names to verifier functions, which is consulted when validating a transaction. A verifier may charge gas after gas is bought and before any transaction code is run. Because of this, currently, a verifier cannot grant the `GAS_PAYER` capability, which would allow paying for gas using a verifier’s attestation. This may change in the future.
+
+Notably, writing a verifier plugin does not require changes to `pact`, only to `chainweb-node`, because verifiers run on the Pact container layer and verifier names are strings.
+
+## Changes to the Pact transaction schema
+
+For reference, this is the schema of the `signers` field:
```

Comment:
Whats the connection between `signers` and `verifiers` field? Is it included here to show that both have similar type structures?

Reply:
Yes that's all.